src/runtime/FinalizationRegistryObject.cpp

```diff
@@ -73,4 +90,96 @@ bool FinalizationRegistryObject::deleteCell(ExecutionState& state, Value unregis
     return removed;
 }
 
+class CleanupSomeJob : public Job {
+public:
+    CleanupSomeJob(Context* relatedContext, FinalizationRegistryObject* object, Optional<Object*> callback)
+        : Job(relatedContext)
+        , m_object(object)
+        , m_callback(callback)
+    {
+    }
+
+    SandBox::SandBoxResult run()
+    {
+        auto oldCallback = m_object->m_cleanupCallback;
+        m_object->m_cleanupCallback = m_callback;
+
+        clearStack<1024>();
+        GC_gcollect_and_unmap();
+        GC_gcollect_and_unmap();
+        GC_gcollect_and_unmap();
+        m_object->m_cleanupCallback = oldCallback;
+
+        SandBox sandbox(relatedContext());
+        ExecutionState state(relatedContext());
+        return sandbox.run([&]() -> Value {
+            return Value();
+        });
+    }
+
+private:
+    FinalizationRegistryObject* m_object;
+    Optional<Object*> m_callback;
+};
+
+void FinalizationRegistryObject::cleanupSome(ExecutionState& state, Optional<Object*> callback)
+{
+    state.context()->vmInstance()->jobQueue()->enqueueJob(new CleanupSomeJob(state.context(), this, callback));
+}
+
+void* FinalizationRegistryObject::FinalizationRegistryObjectItem::operator new(size_t size)
+{
+#ifdef NDEBUG
+    static bool typeInited = false;
+    static GC_descr descr;
+    if (!typeInited) {
+        GC_word obj_bitmap[GC_BITMAP_SIZE(FinalizationRegistryObject::FinalizationRegistryObjectItem)] = { 0 };
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject::FinalizationRegistryObjectItem, source));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject::FinalizationRegistryObjectItem, heldValue));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject::FinalizationRegistryObjectItem, unregisterToken));
+        descr = GC_make_descriptor(obj_bitmap, GC_WORD_LEN(FinalizationRegistryObject::FinalizationRegistryObjectItem));
+        typeInited = true;
+    }
+    return GC_MALLOC_EXPLICITLY_TYPED(size, descr);
+#else
+    return CustomAllocator<FinalizationRegistryObjectItem>().allocate(1);
+#endif
+}
+
+void FinalizationRegistryObject::finalizer(Object* self, void* data)
+{
+    FinalizationRegistryObjectItem* item = (FinalizationRegistryObjectItem*)data;
+    // Vector::erase call should not trigger GC
+    GC_disable();
+
+    auto copyedCells = item->source->m_cells;
+    for (size_t i = 0; i < item->source->m_cells.size(); i++) {
+        if (self == item->source->m_cells[i]->weakRefTarget) {
+            item->source->m_cells.erase(i);
+            i--;
+        }
+    }
+
+    GC_enable();
```

Comment:
I have some questions about this part.
First, why do you block `gc` operation during this part? nested finalizer calls might be a problem?
And copying cells is really necessary?

Reply:
1. I want to avoid nested finalizer calls. IMO nested finalizer calls might trigger unexpected behavior
(If I cannot find the case even multi-thread enabled, I will remove block)
2. copying cells is necessary. because finalizer callback can modify m_cells.

Reply:
I got it :)